azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```
@@ -0,0 +1,151 @@
+
+
+class AsyncHandlerManagerException(ChainableException):
+    pass
+
+
+class AsyncHandlerManager(object):
+    def __init__(self, inbox_manager):
+        self._inbox_manager = inbox_manager
+
+        self._handler_tasks = {
+            # Inbox handler tasks
+            MESSAGE: None,
+            METHOD: None,
+            TWIN_DP_PATCH: None,
+            # Other handler tasks
+            # TODO: add
+        }
+
+        # Inbox handlers
+        self._on_message_received = None
+        self._on_method_request_received = None
+        self._on_twin_desired_properties_patch_received = None
+
+        # Other handlers
+        # TODO: add
+
+    async def _run_inbox_handler(self, inbox, handler_name):
+        """Run infinite loop that waits for an inbox to receive an object from it, then calls
+        the handler with that object
+        """
+        # Run the handler in a threadpool, so that it cannot block other handlers (from a different task),
+        # or the main client thread. The number of worker threads forms an upper bound on how many instances
+        # of the same handler can be running simultaneously.
+        tp = ThreadPoolExecutor(max_workers=4)
+        while True:
+            retval = await inbox.get()
+            # NOTE: we MUST use getattr here using the handler name, as opposed to directly passing
+            # the handler in order for the handler to be able to be updated without cancelling
+            # the running task created for this coroutine
+            handler = getattr(self, handler_name)
+            if inspect.iscoroutinefunction(handler):
+                # Wrap the coroutine in a function so it can be run in ThreadPool
+                def coro_wrapper(coro, arg):
+                    asyncio_compat.run(coro(arg))
+
+                tp.submit(coro_wrapper, handler, retval)
+            else:
+                # Run function directly in ThreadPool
+                tp.submit(handler, retval)
+
```

same comment from 4 lines up on exceptions being swallowed.